server/src/main/java/io/crate/execution/engine/collect/CollectTask.java

```diff
@@ -60,16 +64,20 @@
     private final SharedShardContexts sharedShardContexts;
 
     private final IntObjectHashMap<RefCountedItem<? extends IndexSearcher>> searchers = new IntObjectHashMap<>();
-    private final Object subContextLock = new Object();
     private final RowConsumer consumer;
     private final int ramAccountingBlockSizeInBytes;
+
+    @GuardedBy("searchers")
     private final ArrayList<MemoryManager> memoryManagers = new ArrayList<>();
     private final Version minNodeVersion;
+    private final CompletableFuture<Void> consumerCompleted;
+    private final CompletableFuture<BatchIterator<Row>> batchIterator = new CompletableFuture<>();
+    private final AtomicBoolean started = new AtomicBoolean(false);
 
-    private BatchIterator<Row> batchIterator = null;
+    private boolean releasedResources = false;
```

Comment:
shouldn't this be volatile as it may be read and written from different threads?

Reply:
It's always within a synchronized block and synchronized causes memory visibility, so the volatile is not necessary. Should I add a `@GuardedBy` annotation?

Reply:
Added the annotation